lib/router/src/cache.rs

```diff
@@ -1,426 +1,397 @@
+                // The cache is at capacity, but we are able to remove a value
+                Ok(Async::Ready(Some(expired))) => {
+                    self.values.remove(expired.get_ref());
+                }
+
+                // `Ready(None)` can only be returned when `expirations` is
+                // empty; we know `expirations` is not empty because `values`
+                // is not empty and capacity does not equal zero.
+                Ok(Async::Ready(None)) => unreachable!("cache expirations cannot be empty"),
+
+                // The cache is at capacity and no values can be removed
+                Ok(Async::NotReady) => {
+                    return Ok(Async::NotReady);
+                }
+
+                Err(e) => panic!("Cache.expirations DelayQueue::poll must not fail: {}", e),
+            }
+        }
 
-impl<'a, K: Hash + Eq + 'a, V: 'a, N: Now + 'a> Reserve<'a, K, V, N> {
-    /// Stores a route in the cache.
-    pub fn store(self, key: K, val: V) {
-        let node = Node::new(val.into(), self.now.now());
-        self.vals.insert(key, node);
+        Ok(Async::Ready(()))
     }
-}
 
-// ===== impl Access =====
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(expired) = try_ready!(self.expirations.poll()) {
+            self.values.remove(expired.get_ref());
+        }
 
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
+        Ok(Async::Ready(()))
     }
 }
 
-impl<'a, T: 'a, N: Now + 'a> DerefMut for Access<'a, T, N> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.node
-    }
-}
+// ===== impl PurgeCache =====
 
-impl<'a, T: 'a, N: Now + 'a> Access<'a, T, N> {
-    #[cfg(test)]
-    fn last_access(&self) -> Instant {
-        self.node.last_access
-    }
-}
+impl<K, V> Future for PurgeCache<K, V>
+where
+    K: Clone + Eq + Hash,
+    V: Clone,
+{
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        try_ready!(self.interval.poll().map_err(|e| {
+            panic!("PurgeCache.interval Interval::poll must not fail: {}", e);
+        }));
+
+        let mut cache = try_ready!(Ok(self.cache.poll_lock()));
 
-impl<'a, T: 'a, N: Now + 'a> Drop for Access<'a, T, N> {
-    fn drop(&mut self) {
-        self.node.last_access = self.now.now();
+        cache.poll_purge().map_err(|e| {
+            panic!("PurgeCache.cache Cache::poll_purge must not fail: {}", e);
+        })
     }
 }
 
 // ===== impl Node =====
 
 impl<T> Node<T> {
-    pub fn new(value: T, last_access: Instant) -> Self {
-        Node { value, last_access }
+    pub fn new(dq_key: delay_queue::Key, value: T) -> Self {
+        Node { dq_key, value }
     }
 
-    pub fn access<'a, N: Now + 'a>(&'a mut self, now: &'a N) -> Access<'a, T, N> {
-        Access { now, node: self }
+    pub fn dq_key_ref(&self) -> &delay_queue::Key {
```

Comment:
I don't really get why this needs to be `pub`. AFAICT, it's not used anywhere outside of the cache module. Actually, now that I think of it, it looks like the `Node` type need not be `pub`?

Reply:
Yep I don't think these even needed to be public in the original implementation. I'll remove them in my next set of changes.

Reply:
Well `Cache::insert` does return an `Option<Node<V>>` so it would need to be pub for that. Field access through the impl doesn't need to be pub for how it is used right now, but maybe something consider in the future? I was basing some of that off [delay_queue::Expired](https://docs.rs/tokio-timer/0.2.10/src/tokio_timer/delay_queue.rs.html#830-845)

Reply:
Ah, I missed that `Cache::insert` returned a `Node`. Is it necessary for it to return `Option<Node<V>>` rather than `Option<V>`? As far as I can tell, the `Node` is never actually used. Can we just map over the removed node and extract the `V`?

Reply:
Yep `Node` is an implementation detail that is only used in _cache.rs_. It makes sense to just return the value instead of keeping it coupled with the `delay_queue::Key`